### ISSUE TITLE
TD004: Support additional TODO comment formats

### ISSUE THREAD

**Issue by solonovamax** [astral-sh/ruff#26451]
### Summary

I think it would be good if additional formats for TODO comments could be supported.

Personally, the format I use is:
```py
# TODO [date] ([user]): [message]
```
for example,
```py
# TODO 2026-06-29 (solonovamax): Fix this error
```

I specifically like this format because the date tells you when the TODO comment was written.

I think it would be nice if additional formats, such as mine, could be supported. Perhaps offering a customizeable regex?

**Comment by VascoSch92** [comment]
I could also think that Ruff expose some kind of regex to set your preferred format. But just an idea.

**Comment by MichaReiser** [comment]
This is a generalization of https://github.com/astral-sh/ruff/issues/20809, also related to https://github.com/astral-sh/ruff/issues/16963, and https://github.com/astral-sh/ruff/issues/11323

**Comment by VascoSch92** [comment]
@MichaReiser Do you want to proceed with this rule?

Actually, I think it's possible to create a mechanism that checks TODO comments against a regex, and all the other rules can just extend from it.

**Comment by MichaReiser** [comment]
I let @ntBre answer this, because he has been reviewing #26002

**Comment by ntBre** [comment]
I like the idea of exposing a regex. It seems like lots of people have slightly different needs, so configuring it directly seems like a good approach.

**Comment by VascoSch92** [comment]
@ntBre If you don't mind I can give a shot on this.

**Comment by ntBre** [comment]
I might need to think about the scope of this setting a bit more. Are you planning to work on a regex version of all the TD rules? It doesn't look super straightforward to drop in a `lint.flake8-todos.regex` setting in the existing implementations.

It seems that something like a bunch of named capture groups could potentially work like:

```
(?<tag>...)\((?<author>...)\) ...
```

but that might be quite an unwieldy regex to have to configure. So it might need a bit of design work, but a prototype may also help that discussion.

**Comment by solonovamax** [comment]
perhaps if it becomes too complicated to offer a regex, a slightly different approach might be viable:
offer sufficient knobs so that users can easily configure what the want

here's a hypothetical config layout:
```toml
[lint.flake8-todos]
todo-tags = ["TODO", "FIXME"]
todo-format = [
  "TODO",
  { type = "date", format = "dd-MM-YYYY" },
  { type = "literal", value = " (" },
  "username",
  { type = "literal", value = "):" },
  "message"
]
message-url = true # message must include a URL (beginning with http:// or https://)
message-issue = true # message must include a GitHub-style issue (#1234)
message-jira = true # message must include a JIRA-style issue (FOO-1234)
```
in this hypothetical syntax, the plain strings are just shorthands for `{ type = "whatever" }` (e.g. `"username" is equivalent to `{ type = "username" }`)

alternatively, instead of having the `message-*` properties, something like `message-regex` which is matched against the message itself could be used (does not need to match the entire message, just that it must match some part of the message), so if I wanted to allow for github issues I would use `message-regex = "#\d+"`

or maybe if that message format seems a little too complicated, something like:
```toml
[lint.flake8-todos]
todo-tags = ["TODO", "FIXME"]
todo-format = "%{todo} %{date:dd-MM-YYYY} (%{username}): %{message}
```

both of these would allow for greater validation than a regex, for example "is this a real date", etc.

just some random thoughts.

**Comment by VascoSch92** [comment]
yes, I was scoping for a regex and it is a little convoluted to use. 

The proposition above seems more user-friendly actually
